clients.go

```diff
@@ -26,24 +34,57 @@ func (client *KustoIngestionClient) SendAsync(msg PubsubMsg) error {
 		log.Fatalf("Failed to set function key env var")
 	}
 
-	requestUri := targetUri + "?code=" + functionKey
+	INGESTIONSERVICEBATCHSIZE := os.Getenv("INGESTIONSERVICEBATCHSIZE")
+	batchSize, err := strconv.Atoi(INGESTIONSERVICEBATCHSIZE)
+	if INGESTIONSERVICEBATCHSIZE == "" || err != nil {
+		log.Fatalf("Failed to set INGESTIONSERVICEBATCHSIZE %v", err)
+	}
+
+	requestUri := targetUri + "?code=" + functionKey;
+
+	pubSubMsgs := make([]PubsubMsg, 0, batchSize)
+
+	return KustoIngestionClient { 
+		requestUri:requestUri,
+		pubSubMsgs: pubSubMsgs,
+		batchSize: batchSize,
+	}
+}
+
+func (client *KustoIngestionClient) SendAsync(msg PubsubMsg) error {
+	client.mutex.Lock()
+	defer client.mutex.Unlock()
+
+	client.pubSubMsgs = append(client.pubSubMsgs, msg);
+	client.counter += 1
+
+	if client.counter == client.batchSize {
+		msgs := make([]PubsubMsg, client.batchSize, client.batchSize)
+		copy(msgs, client.pubSubMsgs)
 
-	rawMsgs := RawMsgs{
-		Records: []PubsubMsg{msg},
+		client.counter = 0
+		client.pubSubMsgs = make([]PubsubMsg, 0, client.batchSize)
+		return client.SendAsyncBatch(msgs)
 	}
```

Comment:
Would the last (incomplete) batch of items be still sent with this logic? Not familiar with Go enough to see if that is the case but wanted to confirm ... :)

Reply:
Since the events are fixed, ie 1M, the 1K batch size should always be published - this is not resilient enough to handle variable but should be fine for testing.

Reply:
the publisher is publishing forever from an infinite loop. So, theoretically, the batch size will always get satisfied. However, if the publisher errors out and stops sending, we could have an edge case in which we will have an incomplete batch. In that case, we will have an estimated drop rate since the subscriber will receive events but not ingest them in Kusto.

Reply:
Sounds good, we can customize batch size later if needed.